File: backend/src/application/action_navigation.py
from __future__ import annotations

from urllib.parse import urlsplit, urlunsplit

DEFAULT_ACTION_HREF = "/record"

VALID_ACTION_ROUTES = frozenset(
    {
        "/analysis",
        "/community",
        "/dashboard",
        "/hospital",
        "/notifications",
        "/record",
        "/schedule",
        "/settings",
        "/shopping",
        "/timeline",
    }
)

ACTION_ROUTE_ALIASES = {
    "/calendar": "/schedule",
    "/hospital-recommendations": "/hospital",
    "/hospitals": "/hospital",
    "/products": "/shopping",
    "/schedules": "/schedule",
    "/shop": "/shopping",
}


def normalize_action_href(href: str | None, fallback: str | None = DEFAULT_ACTION_HREF) -> str | None:
    if href is None:
        return fallback

    stripped = href.strip()
    if not stripped or not stripped.startswith("/") or stripped.startswith("//"):
        return fallback

    parsed = urlsplit(stripped)
    if parsed.scheme or parsed.netloc:
        return fallback

    path = ACTION_ROUTE_ALIASES.get(parsed.path, parsed.path)
    if path not in VALID_ACTION_ROUTES:
        return fallback

    return urlunsplit(("", "", path, parsed.query, parsed.fragment))

File: frontend/app/web/src/lib/action-navigation.ts
const routeAliases: Record<string, string> = {
  "/calendar": "/schedule",
  "/hospital-recommendations": "/hospital",
  "/hospitals": "/hospital",
  "/products": "/shopping",
  "/schedules": "/schedule",
  "/shop": "/shopping",
};

const validCareRoutes = new Set([
  "/",
  "/analysis",
  "/community",
  "/hospital",
  "/more",
  "/notifications",
  "/profile",
  "/record",
  "/schedule",
  "/settings",
  "/shared-care",
  "/shopping",
  "/suggestions",
  "/timeline",
]);

type CareActionSource = {
  action?: string;
  actionHref?: string | null;
  category?: string;
  detail?: string;
  reason?: string;
  title?: string;
};

export function normalizeCareActionHref(href: string | null | undefined, fallbackHref = "/record") {
  const trimmedHref = href?.trim();
  if (!trimmedHref || !trimmedHref.startsWith("/") || trimmedHref.startsWith("//")) {
    return fallbackHref;
  }

  const [, pathname = trimmedHref, suffix = ""] = trimmedHref.match(/^([^?#]*)([?#].*)?$/) ?? [];
  const normalizedPathname = routeAliases[pathname] ?? pathname;

  if (!validCareRoutes.has(normalizedPathname)) {
    return fallbackHref;
  }

  return `${normalizedPathname}${suffix}`;
}

export function getCareActionHref(actionHref: string | null | undefined, fallbackHref = "/record") {
  return normalizeCareActionHref(actionHref, fallbackHref);
}

export function getCareSuggestionHref(suggestion: CareActionSource, fallbackHref = "/record") {
  return getCareActionHref(suggestion.actionHref, fallbackHref);
}

export function getCareNotificationHref(notification: CareActionSource, fallbackHref = "/record") {
  return getCareActionHref(notification.actionHref, fallbackHref);
}
